### PR TITLE
NCNE changes for adding work flow types for normal chart types and th…

### PR DESCRIPTION
**Pull request by ssrajan**
NCNE changes to create workflow steps 
1. StageType factory to get different stages for different task types
2. Test cases for StageType factory
3. Post deployment script to insert standard taskstagetypes to the table.

### CODE REVIEW

File: src/NCNEPortal/Helpers/StageTypeFactory.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    public class StageTypeFactory : IStageTypeFactory
+    {
+        private readonly NcneWorkflowDbContext _dbContext;
+
+        public StageTypeFactory(NcneWorkflowDbContext dbContext)
+        {
+            _dbContext = dbContext;
+        }
+
+        public List<TaskStageType> GeTaskStages(string chartType)
```
Comment (benbhall) [inline src/NCNEPortal/Helpers/StageTypeFactory.cs]:
```suggestion
        public List<TaskStageType> GetTaskStages(string chartType)
```

File: src/NCNEPortal.UnitTests/WorkflowTests.cs [hunk 2]
```diff
@@ -242,6 +252,41 @@ public async Task Test_3ps_Entering_Actual_return_date_without_expected_return_d
...
+
+            Assert.AreEqual(stagetypes.Count, 18);
+            Assert.AreEqual(Sdra.Count(), 1);
+        }
+
+
+        [Test]
+        public async Task Test_StageType_Factory_For_Primary_ChartType()
+        {
+            var stagetypes = _stageTypeFactory.GeTaskStages("Primary");
```
Comment (benbhall) [inline src/NCNEPortal.UnitTests/WorkflowTests.cs]:
```suggestion
            var stagetypes = _stageTypeFactory.GetTaskStages("Primary");
```

File: src/NCNEPortal.UnitTests/WorkflowTests.cs [hunk 3]
```diff
@@ -242,6 +252,41 @@ public async Task Test_3ps_Entering_Actual_return_date_without_expected_return_d
             Assert.IsTrue(_workflowModel.ValidationErrorMessages.Contains("3PS : Please enter expected return date before entering actual return date"));
         }
 
+        [Test]
+        public async Task Test_StageType_Factory_For_Adoption_ChartType()
+        {
+            var stagetypes = _stageTypeFactory.GeTaskStages("Adoption");
```
Comment (benbhall) [inline src/NCNEPortal.UnitTests/WorkflowTests.cs]:
```suggestion
            var stagetypes = _stageTypeFactory.GetTaskStages("Adoption");
```

File: src/NCNEPortal/Pages/NewTask.cshtml.cs [hunk 4]
```diff
@@ -208,6 +175,123 @@ public async Task<IActionResult> OnPost()
...
+
+            _logger.LogInformation($"Deadline Dates recalculated before saving");
+
+        }
+
+        private async Task CreateTaskStages(int processId)
+        {
+
+
+            foreach (var taskStageType in _stageTypeFactory.GeTaskStages(ChartType))
```
Comment (benbhall) [inline src/NCNEPortal/Pages/NewTask.cshtml.cs]:
```suggestion
            foreach (var taskStageType in _stageTypeFactory.GetTaskStages(ChartType))
```

File: src/NCNEPortal/Pages/NewTask.cshtml.cs [hunk 5]
```diff
@@ -208,6 +175,123 @@ public async Task<IActionResult> OnPost()
...
+            {
+                var taskStage = _ncneWorkflowDbContext.TaskStage.Add(new TaskStage()).Entity;
+
+                taskStage.ProcessId = processId;
+                taskStage.TaskStageTypeId = taskStageType.TaskStageTypeId;
+
+                //Assign the status of the task stage 
+                taskStage.Status = (NcneTaskStageType)taskStageType.TaskStageTypeId switch
```
Comment (benbhall) [inline src/NCNEPortal/Pages/NewTask.cshtml.cs]: Great to see some C# 8 @ssrajan! :)

File: src/NCNEPortal/Pages/NewTask.cshtml.cs [hunk 6]
```diff
@@ -208,6 +175,123 @@ public async Task<IActionResult> OnPost()
...
+                        (DateTime)this.PublicationDate);
+
+                this.CommitToPrintDate = commitDate;
+                this.CISDate = cisDate;
+                this.AnnounceDate = formsDate;
+
+
+            }
+
+            _logger.LogInformation($"Deadline Dates recalculated before saving");
```
Comment (benbhall) [inline src/NCNEPortal/Pages/NewTask.cshtml.cs]: Is it right to be logging that dates were recalculated? There is no `else` to handle a situation where `RepromatDate` is null, thus `PulicationDate` is null. I think we're swallowing a possible null reference without logging it or handling it in any other way?

File: src/NCNEPortal/Helpers/StageTypeFactory.cs (hunk 1 above)
Comment (ssrajan) [reply src/NCNEPortal/Helpers/StageTypeFactory.cs]: Corrected

File: src/NCNEPortal/Pages/NewTask.cshtml.cs (hunk 5 above)
Comment (ssrajan) [reply src/NCNEPortal/Pages/NewTask.cshtml.cs]: Thank you

File: src/NCNEPortal/Pages/NewTask.cshtml.cs (hunk 6 above)
Comment (ssrajan) [reply src/NCNEPortal/Pages/NewTask.cshtml.cs]: That recalculation itself is to over-come the postback not sending disabled controlls. Null is allowed in all dates and they won't be recalculated if the source date is null. So, the log is unnecessary there. I have removed it.

**Pull request merged**